processes/processes.tex

```diff
@@ -2,196 +2,229 @@ \chapter{Processes}
+As most of you have realized by now, if you stuck the following code snippet into a program, the variables would not be shared between two parallel invocations of the program.
 
 \begin{lstlisting}[language=C]
 int secrets;
 secrets++;
 printf("%d\n", secrets);
 \end{lstlisting}
 
-On two different terminals, as you would guess they would both print out 1 not 2.
-Even if we changed the code to do something really hacky, there would be no way to change another process' state unintentionally.
-There are ways of changing it purposefully though.
+On two different terminals, they would both print out 1 not 2.
+Even if we changed the code to attempt to affect other process instances, there would be no way to change another process' state unintentionally.
+However, there are other intentional ways to change the program states of other processes.
 
 \section{Process Contents}
 
 \subsection{Memory Layout}
 
-When a process starts, it gets its own address space. Each process gets the following.
+When a process starts, it gets its own address space. Each process gets the following:
 
 \begin{itemize}
 \item \textbf{A Stack}.
-The stack is the place where automatic variable and function call return addresses are stored.
+The stack is the place where automatically allocated variables and function call return addresses are stored.
 Every time a new variable is declared, the program moves the stack pointer down to reserve space for the variable.
-This segment of the stack is Writable but not executable.
-If the stack grows too far -- meaning that it either grows beyond a preset boundary or intersects the heap -- you will get a stackoverflow most likely resulting in a SEGFAULT.
-\textbf{The stack is statically allocated by default meaning that there is only a certain amount of space to which one can write}
+This segment of the stack is writable but not executable (this behavior is controlled by the no-execute(NX) bit).
+If the stack grows too far -- meaning that it either grows beyond a preset boundary or intersects the heap -- you will get a stack overflow error, most likely resulting in a SEGFAULT.
+\textbf{The stack is statically allocated by default; there is only a certain amount of space to which one can write.}
 
 \item \textbf{A Heap}.
 The heap is a contiguous, expanding region of memory \cite{mallocinternals}.
-If you want to allocate a large object, it goes here.
-The heap starts at the top of the text segment and grows upward, meaning sometimes when you call \keyword{malloc} that it asks the operating system to push the heap boundary upward.
-More on that in the memory allocation chapter.
+If you want to allocate a object whose memory you want to manually control, or whose size cannot be determined at compile time, you would want to create a heap variable.
```

Comment:
We should probably specify what "manually" means

* If you want to allocate an object whose memory is still valid after a function call,